### PR TITLE
VR: Fix Redundant VRouter guest network on wrong interface

### PR DESCRIPTION
**Pull request by weizhouapache**
## Description

This PR aims to fix the issues on KVM environments with linux bridge and OVS if there are multiple public IPs associated to a network or vpc.

fixes: #3179 

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

- [X] Tested the following network on KVM with linux bridge and KVM with OVS.

- networks with single VR
- networks with redundant VRs

Actions
  1. Create network offering with single VR, and enable it
  2. create isolated network with the network offering
  3. create a vm in the network.
    verify the available nics in VR should be "eth0,eth1,eth2"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP
  4. get a free public ip, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP and new ip
  5. remove the port forwarding rule, and release the new ip
    verify the available nics in VR should be "eth0,eth1,eth2"
    verify the IPs in VR. eth0 -> guest nic IP, eth2 -> source nat IP
 
  6. create new public ip range 1
  7. get a free ip in new ip range, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2,eth3"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 1
  8. get a free ip in new ip range, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2,eth3"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 1, new ip 2,
  9. get a free ip in new ip range, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2,eth3"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 1, new ip 2, new ip 3
  10. release new ip 2
    verify the available nics in VR should be "eth0,eth1,eth2,eth3"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 1, new ip 3
  11. release new ip 1
    verify the available nics in VR should be "eth0,eth1,eth2,eth3"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 3
  12. create new public ip range 2
  13. get a free ip 4 in new ip range 2, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 3, eth4 -> new ip 4
  14. get a free ip 5 in new ip range 2, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 3, eth4 -> new ip 4/5
  15. get a free ip 6 in new ip range 2, assign to network, and create port forwarding rules (ssh) to the vm
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 3, eth4 -> new ip 4/5/6
  16. release new ip 5
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 3, eth4 -> new ip 4/6
  17. release new ip 4
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 3, eth4 -> new ip 6
  18. release new ip 3
    verify the available nics in VR should be "eth0,eth1,eth2,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth4 -> new ip 6
  19. restart network
    verify the available nics in VR should be "eth0,eth1,eth2,eth4,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth4 -> new ip 6
  20. restart network with cleanup
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 6
  21. restart network with cleanup, makeredundant=true
    verify the available nics in VR should be "eth0,eth1,eth2,eth3,"
    verify the IPs in VR. eth0 -> guest nic, eth2 -> source nat IP, eth3 -> new ip 6

- [X] Tested the following vpcs on KVM with linux bridge and KVM with OVS.

- vpc with single VR
- vpc with redundant VRs

Actions
 1. get vpc offering with single VR
 2. create a vpc with the vpc offering
   verify the available nics in VR should be "eth0,eth1"
   verify the IPs in VR. eth0 -> control nic, eth1 -> source nat IP
 3. create a tier in the vpc, and create a vm in the tier.
   verify the available nics in VR should be "eth0,eth1,eth2"
   verify the IPs in VR. eth0 -> control nic, eth1 -> source nat IP, eth2 -> tier 1
 4. get a free public ip, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2"
   verify the IPs in VR. eth0 -> control nic, eth1 -> source nat IP, eth2 -> tier 1
 5. remove the port forwarding rule, and release the new ip
   verify the available nics in VR should be "eth0,eth1,eth2"
   verify the IPs in VR. eth0 -> control nic, eth1 -> source nat IP, eth2 -> tier 1

 6. create new public ip range 1
 7. get a free ip in new ip range, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2,eth3"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 1
 8. get a free ip in new ip range, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2,eth3"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 1/2
 9. get a free ip in new ip range, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2,eth3"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 1/2/3
 10. release new ip 2
   verify the available nics in VR should be "eth0,eth1,eth2,eth3"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 1/3
 11. release new ip 1
   verify the available nics in VR should be "eth0,eth1,eth2,eth3"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3
 12. create a tier 2 in the vpc, and create a vm 2 in the tier2.
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3, eth4 -> tier 2

 13. create new public ip range 2
 14. get a free ip 4 in new ip range 2, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3, eth4 -> tier 2, eth5 -> new ip 4
 15. get a free ip 5 in new ip range 2, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3, eth4 -> tier 2, eth5 -> new ip 4/5
 16. get a free ip 6 in new ip range 2, assign to network, and create port forwarding rules (ssh) to the vm
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3, eth4 -> tier 2, eth5 -> new ip 4/5/6
 17. release new ip 5
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3, eth4 -> tier 2, eth5 -> new ip 4/6
 18. release new ip 4
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth3 -> new ip 3, eth4 -> tier 2, eth5 -> new ip 6
 19. release new ip 3
   verify the available nics in VR should be "eth0,eth1,eth2,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth4 -> tier 2, eth5 -> new ip 6
 20. restart tier1
   verify the available nics in VR should be "eth0,eth1,eth2,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth4 -> tier 2, eth5 -> new ip 6
 20. restart tier2
   verify the available nics in VR should be "eth0,eth1,eth2,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth4 -> tier 2, eth5 -> new ip 6
 22. restart VPC
   verify the available nics in VR should be "eth0,eth1,eth2,eth4,eth5,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> tier 1, eth4 -> tier 2, eth5 -> new ip 6
 23. restart VPC with cleanup
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> new ip 6, eth3 -> tier 1, eth4 -> tier 2
 24. restart VPC with cleanup, makeredundant=true
   verify the available nics in VR should be "eth0,eth1,eth2,eth3,eth4,"
   verify the IPs in VR. eth1 -> source nat IP, eth2 -> new ip 6, eth3 -> tier 1, eth4 -> tier 2

### CODE REVIEW

**Comment by weizhouapache** [comment]
@svenvogel @DennisKonrad 
rebased with latest master. please test it.

**Comment by weizhouapache** [comment]
@DaanHoogland @rhtyd could you please kick travis test ?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by svenvogel** [comment]
@weizhouapache thanks. we will test it and come back fastly.

File: server/src/main/java/com/cloud/network/router/CommandSetupHelper.java [hunk 1]
```diff
@@ -670,7 +670,20 @@ public void createVpcAssociatePublicIPCommands(final VirtualRouter router, final
...
+        for (final NicVO nic : nics) {
+            final NetworkVO nw = _networkDao.findById(nic.getNetworkId());
+            if (nw.getTrafficType() == TrafficType.Guest) {
+                guestNetworkId = nw.getId();
+                break;
+            }
+        }
+
+        Map<String, Boolean> vlanLastIp = getVlanLastIpMap(router.getVpcId(), guestNetworkId);
+
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/router/CommandSetupHelper.java]: could this be a aeparate method? getVlanLastIpMapForGuestNetwork();?

and than later

File: server/src/main/java/com/cloud/network/router/CommandSetupHelper.java [hunk 2]
```diff
@@ -728,6 +741,13 @@ public int compare(final PublicIpAddress o1, final PublicIpAddress o2) {
             final DataCenterVO dcVo = _dcDao.findById(router.getDataCenterId());
             cmd.setAccessDetail(NetworkElementCommand.ZONE_NETWORK_TYPE, dcVo.getNetworkType().toString());
 
+            Boolean lastIp = vlanLastIp.get(vlanTagKey);
+            if (lastIp == null) {
+                cmd.setAccessDetail(NetworkElementCommand.NETWORK_PUB_LAST_IP, "true");
+            } else {
+                cmd.setAccessDetail(NetworkElementCommand.NETWORK_PUB_LAST_IP, "false");
+            }
+
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/router/CommandSetupHelper.java]: here useAccessDetailsAccordingToVlanLastIpForGuestNetwork(); (long names just to be descriptive

**Comment by rhtyd** [comment]
I'll wait for your testing @svenvogel and then review/test. Thanks.

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 3]
```diff
@@ -1794,35 +1794,26 @@ public ExecutionResult prepareNetworkElementCommand(final IpAssocCommand cmd) {
...
-                }
-                nicPos++;
+            conn = getLibvirtUtilitiesHelper().getConnectionByVmName(routerName);
+            Integer devNum = 0;
+            final List<InterfaceDef> pluggedNics = getInterfaces(conn, routerName);
+            final Map<String, Integer> macAddressToNicNum = new HashMap<>(pluggedNics.size());
+            for (final InterfaceDef pluggedNic : pluggedNics) {
+                final String pluggedVlan = pluggedNic.getBrName();
+                macAddressToNicNum.put(pluggedNic.getMacAddress(), devNum);
+                devNum++;
```
Comment (GabrielBrascher) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java]: Duplicated lines (_1797 - 1805_ **=** _1838 - 1847_). Can you please extract them into a method?

File: systemvm/debian/etc/sysctl.conf [hunk 4]
```diff
@@ -27,6 +27,10 @@ net.ipv4.conf.default.send_redirects = 0
 net.ipv4.conf.all.secure_redirects = 0
 net.ipv4.conf.default.secure_redirects = 0
 
+# Prmote secondary ip to be primary if primary IP is removed
```
Comment (GabrielBrascher) [inline systemvm/debian/etc/sysctl.conf]: There is a typo at line 30

**Comment by DaanHoogland** [comment]
@weizhouapache , as the issue (#3179) is marked for 4.13.1 can you rebase, please?

**Comment by DennisKonrad** [comment]
@weizhouapache @rhtyd I'm testing this since last week and as of now the PR looks very solid. I'll come back in some days and add the review when I'm done

**Comment by weizhouapache** [comment]
@DaanHoogland @GabrielBrascher thanks for review.
I will rebase with 4.13 and test it.

**Comment by weizhouapache** [comment]
@DaanHoogland @rhtyd rebased with 4.13
can you kick a test ? Thanks

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@DaanHoogland considering this PR has changes in server project, it would be better to test on xenserver as well .

I have only tested on kvm.

**Comment by DaanHoogland** [comment]
tnx for the heads-up @weizhouapache , our lab is stretched at the moment so i'll have to decide on some priorities but i will start a xen test soon.

**Comment by andrijapanicsb** [comment]
> @weizhouapache @rhtyd I'm testing this since last week and as of now the PR looks very solid. I'll come back in some days and add the review when I'm done

@DennisKonrad @svenvogel - any progress with testing this one?

**Comment by DaanHoogland** [comment]
@weizhouapache, I had a quick look at testing this in xen. For my understanding on how this must be tested, when you say
6. create new public ip range 1
do you mean `createPortableIpRange` or do you mean `createVlanIpRange`?
tnx

**Comment by weizhouapache** [comment]
> @weizhouapache, I had a quick look at testing this in xen. For my understanding on how this must be tested, when you say
> 6. create new public ip range 1
> do you mean `createPortableIpRange` or do you mean `createVlanIpRange`?
> tnx

@DaanHoogland I tested with `createVlanIpRange`

**Comment by DaanHoogland** [comment]
> @DaanHoogland I tested with `createVlanIpRange`

tnx, i'll find some time to test, soon

**Comment by svenvogel** [comment]
@andrijapanicsb @weizhouapache we are one testing. we @DennisKonrad come back in the next two days. 👍

**Comment by DaanHoogland** [comment]
great @svenvogel What kind of platform(s) will you be testing?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DennisKonrad** [comment]
@weizhouapache @DaanHoogland @andrijapanicsb I tested a lot with OvS on KVM. This looks very good.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Env issue, will rekick
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 xenserver74

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by DaanHoogland** [comment]
@blueorangutan sorry

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
We probably want to run test matrix on this
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by andrijapanicsb** [comment]
@svenvogel did you actually test this PR before giving LGTM (in the sense that you said you will be testing it)
@DennisKonrad - what's the progress on your side - can we put some effort on testing is please (I recall this fixes one of the issues you've raised back in 2019)

thx

**Comment by rhtyd** [comment]
Please review, test @andrijapanicsb @DaanHoogland 
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@weizhouapache @andrijapanicsb @rhtyd @DennisKonrad I did not have a chance to test xenserver and won't this week. Are we ok, to merge this as is? regression on xenserver shows ok ^^.

**Comment by svenvogel** [comment]
@andrijapanicsb<https://github.com/andrijapanicsb> yes This PR was tested „ before“ we gave a LGTM. So ... from our side all things on KVM are ok.

From xenserver it should look okay if they was not so much changes but we should ask Wei. @weizhouapache<https://github.com/weizhouapache> what do you think?




__

Sven Vogel
Teamlead Platform

EWERK DIGITAL GmbH
Brühl 24, D-04109 Leipzig
P +49 341 42649 - 99
F +49 341 42649 - 98
S.Vogel@ewerk.com
www.ewerk.com

Geschäftsführer:
Dr. Erik Wende, Hendrik Schubert, Frank Richter
Registergericht: Leipzig HRB 9065

Support:
+49 341 42649 555

Zertifiziert nach:
ISO/IEC 27001:2013
DIN EN ISO 9001:2015
DIN ISO/IEC 20000-1:2011

ISAE 3402 Typ II Assessed

EWERK-Blog<https://blog.ewerk.com/> | LinkedIn<https://www.linkedin.com/company/ewerk-group> | Xing<https://www.xing.com/company/ewerk> | Twitter<https://twitter.com/EWERK_Group> | Facebook<https://de-de.facebook.com/EWERK.IT/>


Mit Handelsregistereintragung vom 09.07.2019 ist die EWERK RZ GmbH auf die EWERK IT GmbH verschmolzen und firmiert nun gemeinsam unter dem Namen: EWERK DIGITAL GmbH, für weitere Informationen klicken Sie hier<https://www.ewerk.com/ewerkdigital>.

Auskünfte und Angebote per Mail sind freibleibend und unverbindlich.

Disclaimer Privacy:
Der Inhalt dieser E-Mail (einschließlich etwaiger beigefügter Dateien) ist vertraulich und nur für den Empfänger bestimmt. Sollten Sie nicht der bestimmungsgemäße Empfänger sein, ist Ihnen jegliche Offenlegung, Vervielfältigung, Weitergabe oder Nutzung des Inhalts untersagt. Bitte informieren Sie in diesem Fall unverzüglich den Absender und löschen Sie die E-Mail (einschließlich etwaiger beigefügter Dateien) von Ihrem System. Vielen Dank.

The contents of this e-mail (including any attachments) are confidential and may be legally privileged. If you are not the intended recipient of this e-mail, any disclosure, copying, distribution or use of its contents is strictly prohibited, and you should please notify the sender immediately and then delete it (including any attachments) from your system. Thank you.

Am 20.02.2020 um 22:28 schrieb dahn <notifications@github.com>:

﻿

@weizhouapache<https://github.com/weizhouapache> @andrijapanicsb<https://github.com/andrijapanicsb> @rhtyd<https://github.com/rhtyd> @DennisKonrad<https://github.com/DennisKonrad> I did not have a chance to test xenserver and won't this week. Are we ok, to merge this as is? regression on xenserver shows ok ^^.

**Comment by DaanHoogland** [comment]
thanks @weizhouapache for the extensive added integration test coverage. You put them in a package called `leaseweb`. Can you explain why? (to a wider audience)

**Comment by weizhouapache** [comment]
> thanks @weizhouapache for the extensive added integration test coverage. You put them in a package called `leaseweb`. Can you explain why? (to a wider audience)

@DaanHoogland moved to test/integration/component/

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
as in https://github.com/apache/cloudstack/pull/3553#issuecomment-592095851 this nic error is new. We might have merged something that causes it, but it disappeared after applying insanity there.
@blueorangutan test

**Comment by DaanHoogland** [comment]
ok, other errors without recompile, merging

**Pull request merged**